File: app/src/main/java/vedam/subkuch/ui/wallet/WalletActivity.java
package vedam.subkuch.ui.wallet;

import android.app.Activity;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.text.Html;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.android.volley.Response;

import java.util.ArrayList;
import java.util.Stack;

import vedam.subkuch.R;
import vedam.subkuch.base.BaseActivity;
import vedam.subkuch.databinding.ActivityWalletBinding;
import vedam.subkuch.helpers.Constants;
import vedam.subkuch.network.DataFetcher;
import vedam.subkuch.network.models.referral.MyReferral;
import vedam.subkuch.network.models.referral.MyReferralResponse;
import vedam.subkuch.network.models.referral.ReferralDetails;
import vedam.subkuch.network.models.wallet.ProfileData;
import vedam.subkuch.network.models.wallet.TermsCondition;
import vedam.subkuch.network.models.wallet.Wallet;
import vedam.subkuch.network.models.wallet.WalletDetails;
import vedam.subkuch.network.models.wallet.WalletResponse;
import vedam.subkuch.utils.AppUtil;
import vedam.subkuch.utils.UiUtil;

public class WalletActivity extends BaseActivity {

    private ActivityWalletBinding activityWalletBinding;
    private Stack<Object> requestStack = new Stack<>();
    private WalletResponse walletResponse;
    private MyReferralResponse myReferralResponse;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityWalletBinding = DataBindingUtil.setContentView(this, R.layout.activity_wallet);
        setTitle(R.string.my_wallet);
        setToolbarBackButton();

        startAPICalls();
    }

    private void startAPICalls() {

        requestStack.add(new Object());
        requestStack.add(new Object());
        UiUtil.showProgressDialog(this, getString(R.string.loading));
        getWalletDetails();
        getMyReferrals();
    }

    private void getWalletDetails() {

        DataFetcher.getWalletDetails(this, onWalletSuccessListener, WalletResponse.class, onErrorListener);
    }

    private void getMyReferrals() {

        DataFetcher.getMyReferral(this, onMyReferralSuccessListener, MyReferralResponse.class, onErrorListener);
    }

    private Response.Listener<WalletResponse> onWalletSuccessListener = response -> {
        requestStack.pop();
        walletResponse = response;
        checkFlagAndLoadUI();
    };

    private Response.Listener<MyReferralResponse> onMyReferralSuccessListener = response -> {
        requestStack.pop();
        myReferralResponse = response;
        checkFlagAndLoadUI();
    };

    private void checkFlagAndLoadUI() {
        if (requestStack.isEmpty()) {
            UiUtil.cancelProgressDialog();
            loadUI();
        }
    }

    private void loadUI() {

        if (walletResponse != null && walletResponse.getReturnMessage().equals(Constants.SUCCESS)) {
            if (walletResponse.getReturnData() != null) {
                bindData(walletResponse.getReturnData());
                hideViews(walletResponse.getReturnData());
            } else {
                UiUtil.showToast(this, getString(R.string.no_data));
                activityWalletBinding.llContainer.setVisibility(View.INVISIBLE);
            }
        } else {
            UiUtil.showToast(this, getString(R.string.err_occurred));
            activityWalletBinding.llContainer.setVisibility(View.INVISIBLE);
        }
    }

    private void bindData(Wallet data) {

        WalletDetails walletDetails = data.getWallet();
        ProfileData profileData = data.getProfileData();
        TermsCondition termsCondition = data.getTermsConditions();
        activityWalletBinding.llContainer.setVisibility(View.VISIBLE);
        UiUtil.setTextViewWithBoldPrefix(this, "Member Name : ",
                AppUtil.getFullName(profileData.getFirstName(), profileData.getLastName()), activityWalletBinding.tvName);
        UiUtil.setTextViewWithBoldPrefix(this, "Phone No. : ", profileData.getMobile(), activityWalletBinding.tvMobile);
        UiUtil.setTextViewWithBoldPrefix(this, "Total Earnings : ", walletDetails.getTotalReferralIncome(), activityWalletBinding.tvTotalEarnings);
        UiUtil.setTextViewWithBoldPrefix(this, "Points Earned : ", walletDetails.getTotalPointsEarned(), activityWalletBinding.tvPointsEarned);
        UiUtil.setTextViewWithBoldPrefix(this, "Withdrawal : ", walletDetails.getTotalWithdrawal(), activityWalletBinding.tvWithdrawal);
        UiUtil.setTextViewWithBoldPrefix(this, "Available Amount : ", walletDetails.getAvailableBalance(), activityWalletBinding.tvAvailableAmount);
        UiUtil.setTextView(activityWalletBinding.tvTncTitle, termsCondition.getTitle());
        setTnc(termsCondition.getDescription());

        boolean isReferralDataAvailable = isReferralDataAvailable();

        if (isReferralDataAvailable) {
            activityWalletBinding.rlSubContainer.setVisibility(View.VISIBLE);
            activityWalletBinding.tvReferralHeading.setVisibility(View.VISIBLE);
            setVenueListener(activityWalletBinding.tvReferee, activityWalletBinding.ivTriangle, activityWalletBinding.rlSubContainer, myReferralResponse.getReturnData());
            setVenue(activityWalletBinding.tvReferee, activityWalletBinding.ivTriangle, myReferralResponse.getReturnData());
        } else {
            activityWalletBinding.rlSubContainer.setVisibility(View.GONE);
            activityWalletBinding.tvReferralHeading.setVisibility(View.GONE);
        }

    }

    private boolean isReferralDataAvailable() {
        return myReferralResponse != null && myReferralResponse.getReturnData() != null
                && myReferralResponse.getReturnData().getReferralDetails() != null
                && !myReferralResponse.getReturnData().getReferralDetails().isEmpty();
    }

    private void setTnc(String description) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N)
            activityWalletBinding.tvTnc.setText(Html.fromHtml(AppUtil.deNull(description), Html.FROM_HTML_MODE_LEGACY));
        else
            activityWalletBinding.tvTnc.setText(Html.fromHtml(AppUtil.deNull(description)));

        activityWalletBinding.tvTnc.setMovementMethod(LinkMovementMethod.getInstance());
    }

    private void setVenueListener(TextView tvVenue, ImageView ivTriangle, RelativeLayout rlSubContainer, ReferralDetails referralDetails) {
        if (referralDetails.getReferralDetails().size() > 2) {
            ivTriangle.setVisibility(View.VISIBLE);
            rlSubContainer.setOnClickListener(v -> {
                if (referralDetails.isExpanded()) {
                    referralDetails.setExpanded(false);
                    setVenue(tvVenue, ivTriangle, referralDetails);
                } else {
                    referralDetails.setExpanded(true);
                    setVenue(tvVenue, ivTriangle, referralDetails);
                }
            });
        } else {
            ivTriangle.setVisibility(View.GONE);
            rlSubContainer.setOnClickListener(null);
        }
    }

    private void setVenue(TextView tvName, ImageView ivTriangle, ReferralDetails referralDetails) {

        if (referralDetails.getReferralDetails().size() > 2)
            if (referralDetails.isExpanded()) {
                tvName.setText(getFullNamesString(referralDetails.getReferralDetails()));
                ivTriangle.setImageResource(R.drawable.baseline_expand_less_black_24dp);
            } else {
                ivTriangle.setImageResource(R.drawable.baseline_expand_more_black_24dp);
                if (referralDetails.getReferralDetails().size() > 0) {
                    MyReferral myReferral = referralDetails.getReferralDetails().get(0);
                    tvName.setText(AppUtil.getFullName(myReferral.getFirstName(), myReferral.getLastName()));
                } else
                    tvName.setText("");
            }
        else {
            tvName.setText(getFullNamesString(referralDetails.getReferralDetails()));
        }
    }

    private String getFullNamesString(ArrayList<MyReferral> myReferrals) {
        StringBuilder fullNames = new StringBuilder();
        for (int i = 0; i < myReferrals.size(); i++) {
            MyReferral myReferral = myReferrals.get(i);
            if (i == myReferrals.size() - 1)
                fullNames.append(AppUtil.getFullName(myReferral.getFirstName(), myReferral.getLastName()));
            else
                fullNames.append(AppUtil.getFullName(myReferral.getFirstName(), myReferral.getLastName())).append("\n");
        }
        return fullNames.toString();
    }

    private void hideViews(Wallet data) {
        WalletDetails walletDetails = data.getWallet();
        ProfileData profileData = data.getProfileData();
        TermsCondition termsCondition = data.getTermsConditions();

        boolean isReferralDataAvailable = isReferralDataAvailable();

        if (TextUtils.isEmpty(profileData.getFirstName()) && TextUtils.isEmpty(profileData.getMobile()) && !isReferralDataAvailable)
            activityWalletBinding.cvName.setVisibility(View.GONE);
        else
            activityWalletBinding.cvName.setVisibility(View.VISIBLE);

        if (TextUtils.isEmpty(walletDetails.getTotalPointsEarned()) && TextUtils.isEmpty(walletDetails.getTotalReferralIncome())
                && TextUtils.isEmpty(walletDetails.getTotalWithdrawal()) && TextUtils.isEmpty(walletDetails.getAvailableBalance()))
            activityWalletBinding.cvEarnings.setVisibility(View.GONE);
        else
            activityWalletBinding.cvEarnings.setVisibility(View.VISIBLE);

        if (TextUtils.isEmpty(termsCondition.getDescription()))
            activityWalletBinding.cvTnc.setVisibility(View.GONE);
        else
            activityWalletBinding.cvTnc.setVisibility(View.VISIBLE);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.withdraw, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_withdraw) {
            startActivityForResult(new Intent(this, WithdrawalActivity.class), Constants.REQUEST_WITHDRAW);
            return true;
        } else
            return super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {

        switch (requestCode) {
            case Constants.REQUEST_WITHDRAW:
                if (resultCode == Activity.RESULT_OK)
                    getWalletDetails();
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: app/src/main/java/vedam/subkuch/ui/wallet/WalletActivity.kt
package vedam.subkuch.ui.wallet

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.result.contract.ActivityResultContracts
import androidx.databinding.DataBindingUtil
import androidx.core.text.HtmlCompat
import androidx.recyclerview.widget.LinearLayoutManager
import vedam.subkuch.network.Response
import vedam.subkuch.R
import vedam.subkuch.base.BaseActivity
import vedam.subkuch.databinding.ActivityWalletBinding
import vedam.subkuch.helpers.Constants
import vedam.subkuch.network.DataFetcher.getMyReferral
import vedam.subkuch.network.DataFetcher.getWalletDetails
import vedam.subkuch.network.DataFetcher.getWalletTerms
import vedam.subkuch.network.models.referral.MyReferral
import vedam.subkuch.network.models.referral.MyReferralResponse
import vedam.subkuch.network.models.wallet.Balance
import vedam.subkuch.network.models.wallet.BalanceResponse
import vedam.subkuch.network.models.wallet.WalletResponse
import vedam.subkuch.utils.AppUtil
import vedam.subkuch.utils.UiUtil
import java.util.*

class WalletActivity : BaseActivity() {
    private var activityWalletBinding: ActivityWalletBinding? = null
    private val requestStack = Stack<Any>()
    private var balanceResponse: BalanceResponse? = null
    private var myReferralResponse: MyReferralResponse? = null
    private var walletTermsResponse: WalletResponse? = null
    private var areReferralsExpanded = false
    private val referralsAdapter = ReferralsAdapter()

    private val withdrawalLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) getWalletDetails()
        }
    private val transferLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) getWalletDetails()
        }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityWalletBinding = DataBindingUtil.setContentView(this, R.layout.activity_wallet)
        setTitle(R.string.my_wallet)
        setToolbarBackButton()
        activityWalletBinding!!.btnTransfer.setOnClickListener { openTransfer() }
        activityWalletBinding!!.btnWithdraw.setOnClickListener { openWithdrawal() }
        activityWalletBinding!!.rvReferrals.apply {
            layoutManager = LinearLayoutManager(this@WalletActivity)
            adapter = referralsAdapter
            isNestedScrollingEnabled = false
            itemAnimator = null
        }
        startAPICalls()
    }

    private fun startAPICalls() {
        requestStack.add(Any())
        requestStack.add(Any())
        UiUtil.showProgressDialog(this, getString(R.string.loading))
        getWalletDetails()
        getMyReferrals()
        getWalletTerms()
    }

    private fun getWalletDetails() {
        getWalletDetails(
            this,
            onBalanceSuccessListener,
            BalanceResponse::class.java,
            onInitialLoadErrorListener
        )
    }

    private fun getMyReferrals() {
        getMyReferral(
            this,
            onMyReferralSuccessListener,
            MyReferralResponse::class.java,
            onInitialLoadErrorListener
        )
    }

    private fun getWalletTerms() {
        getWalletTerms(
            this,
            onWalletTermsSuccessListener,
            WalletResponse::class.java,
            Response.ErrorListener {
                walletTermsResponse = null
                bindTerms()
            }
        )
    }

    private val onBalanceSuccessListener = Response.Listener { response: BalanceResponse? ->
        if (!requestStack.isEmpty()) requestStack.pop()
        balanceResponse = response
        checkFlagAndLoadUI()
    }

    private val onMyReferralSuccessListener = Response.Listener { response: MyReferralResponse? ->
        if (!requestStack.isEmpty()) requestStack.pop()
        myReferralResponse = response
        // A fresh response must start on the first page. Otherwise a refresh after the
        // list has been expanded can leave the pagination control in the wrong state.
        areReferralsExpanded = false
        checkFlagAndLoadUI()
    }

    private val onWalletTermsSuccessListener = Response.Listener { response: WalletResponse? ->
        walletTermsResponse = response
        bindTerms()
    }

    private val onInitialLoadErrorListener = Response.ErrorListener { error ->
        if (!requestStack.isEmpty()) requestStack.pop()
        onErrorReceived(error)
        checkFlagAndLoadUI()
    }

    private fun checkFlagAndLoadUI() {
        if (requestStack.isEmpty()) {
            UiUtil.cancelProgressDialog()
            loadUI()
        }
    }

    private fun loadUI() {
        if (balanceResponse != null && balanceResponse!!.returnMessage == Constants.SUCCESS) {
            if (balanceResponse!!.returnData != null) {
                bindData(balanceResponse!!.returnData)
            } else {
                UiUtil.showToast(this, getString(R.string.no_data))
                activityWalletBinding!!.llContainer.visibility = View.INVISIBLE
            }
        } else {
            UiUtil.showToast(this, getString(R.string.err_occurred))
            activityWalletBinding!!.llContainer.visibility = View.INVISIBLE
        }
    }

    private fun bindData(data: Balance) {
        activityWalletBinding!!.llContainer.visibility = View.VISIBLE
        activityWalletBinding!!.tvTotalEarnings.text = formatCoinValue(data.totalCoinBalance)
        activityWalletBinding!!.tvPointsEarned.visibility = View.GONE
        activityWalletBinding!!.tvWithdrawal.text = formatCoinValue(data.alreadyWithdrawnAmount)
        activityWalletBinding!!.tvAvailableAmount.text =
            formatCoinValue(data.remainingWithdrawableAmount)
        bindReferrals()
        bindTerms()
        activityWalletBinding!!.cvEarnings.visibility = View.VISIBLE
    }

    private fun bindTerms() {
        val binding = activityWalletBinding ?: return
        val terms = walletTermsResponse?.returnData?.termsConditions
        val description = terms?.description?.takeIf { it.isNotBlank() }
        if (description == null) {
            binding.cvTnc.visibility = View.GONE
            return
        }

        binding.tvTncTitle.text = terms.title?.takeIf { it.isNotBlank() }
            ?: getString(R.string.terms_conditions)
        binding.tvTnc.text = HtmlCompat.fromHtml(description, HtmlCompat.FROM_HTML_MODE_LEGACY)
        binding.cvTnc.visibility = View.VISIBLE
    }

    private fun formatCoinValue(value: CharSequence?): String {
        val rawAmount = AppUtil.deNull(value).trim()
        if (rawAmount.isEmpty()) return "0"
        return rawAmount.replaceFirst(
            Regex("^(Vedam Coins|Rs\\.?|NPR|INR|₹|\\$)\\s*", RegexOption.IGNORE_CASE),
            ""
        ).trim()
            // Coin amounts are whole numbers — print "0", not "0.0".
            .replaceFirst(Regex("\\.0+$"), "")
    }

    private fun bindReferrals() {
        val binding = activityWalletBinding ?: return
        val referrals = myReferralResponse
            ?.takeIf { it.returnMessage == Constants.SUCCESS }
            ?.returnData
            ?.referralDetails

        if (referrals == null) {
            referralsAdapter.submitList(emptyList())
            binding.cvName.visibility = View.GONE
            return
        }

        binding.cvName.visibility = View.VISIBLE
        binding.tvMemberNameLabel.visibility = View.GONE
        binding.tvName.visibility = View.GONE
        binding.tvReferralCodeLabel.visibility = View.GONE
        binding.tvMobile.visibility = View.GONE
        binding.referralDivider.visibility = View.GONE
        binding.tvReferralHeading.visibility = View.GONE
        binding.rlSubContainer.visibility = View.VISIBLE
        binding.rlSubContainer.setOnClickListener(null)
        binding.llReferralToggle.setOnClickListener(null)

        if (referrals.isEmpty()) {
            referralsAdapter.submitList(emptyList())
            binding.rvReferrals.visibility = View.GONE
            binding.tvNoReferrals.visibility = View.VISIBLE
            binding.llReferralToggle.visibility = View.GONE
            return
        }

        binding.tvNoReferrals.visibility = View.GONE
        binding.rvReferrals.visibility = View.VISIBLE
        renderReferralNames(referrals)
    }

    private fun renderReferralNames(referrals: List<MyReferral>) {
        val binding = activityWalletBinding ?: return
        val canExpand = referrals.size > MAX_COLLAPSED_REFERRALS
        val visibleReferrals =
            if (canExpand && !areReferralsExpanded) referrals.take(MAX_COLLAPSED_REFERRALS) else referrals
        referralsAdapter.submitList(visibleReferrals)

        if (!canExpand) {
            binding.llReferralToggle.visibility = View.GONE
            binding.llReferralToggle.setOnClickListener(null)
            return
        }

        binding.llReferralToggle.visibility = View.VISIBLE
        binding.ivTriangle.setImageResource(
            if (areReferralsExpanded) {
                R.drawable.baseline_expand_less_black_24dp
            } else {
                R.drawable.baseline_expand_more_black_24dp
            }
        )
        binding.ivTriangle.contentDescription = getString(
            if (areReferralsExpanded) R.string.collapse_referrals else R.string.expand_referrals
        )
        binding.tvReferralToggle.text = if (areReferralsExpanded) {
            getString(R.string.show_fewer_referrals)
        } else {
            getString(R.string.show_all_referrals, referrals.size)
        }
        // Keep pagination on its explicit control. RecyclerView consumes its own touch
        // events, so putting this listener on the surrounding container makes expanding
        // unreliable when the list is tapped.
        binding.llReferralToggle.setOnClickListener {
            areReferralsExpanded = !areReferralsExpanded
            renderReferralNames(referrals)
        }
    }

    private fun openWithdrawal() {
        val intent = Intent(this, WithdrawalActivity::class.java).apply {
            putExtra(
                WithdrawalActivity.EXTRA_AVAILABLE_BALANCE,
                activityWalletBinding!!.tvAvailableAmount.text.toString()
            )
        }
        withdrawalLauncher.launch(intent)
    }

    private fun openTransfer() {
        val intent = Intent(this, TransferActivity::class.java).apply {
            putExtra(
                TransferActivity.EXTRA_AVAILABLE_BALANCE,
                activityWalletBinding!!.tvAvailableAmount.text.toString()
            )
        }
        transferLauncher.launch(intent)
    }

    private companion object {
        const val MAX_COLLAPSED_REFERRALS = 4
    }
}
